src/lib/core/fiber.c

```diff
@@ -1879,6 +1883,7 @@ fiber_init(int (*invoke)(fiber_func f, va_list ap))
 void
 fiber_free(void)
 {
+	trigger_free_in_thread();
 	cord_destroy(&main_cord);
```

Comment:
`cord_destroy` is also called by `coio_on_stop`. AFAIU `trigger_free_in_thread` should be called there, as well. Would be nice if you could put it inside `cord_destroy`.

Reply:
Missed that one, thanks!
I'd love to keep `trigger_init/free_in_thread` in `cord_create` and `cord_destroy`, but `cord_destroy` may be called from `cord_join`, which's in wrong thread. I could make `run_pool` part of `cord` or at least return a pointer to `run_pool` from `trigger_init_in_thread`, and then pass it to `trigger_free_in_thread`. But this doesn't look right to me.

So here we are now. Tell me if you want something different.
<details><summary>diff</summary>

```diff
diff --git a/src/lib/core/coio_task.c b/src/lib/core/coio_task.c
index bae6acc9e..4a2f74a62 100644
--- a/src/lib/core/coio_task.c
+++ b/src/lib/core/coio_task.c
@@ -39,6 +39,7 @@
 #include <sys/socket.h>
 
 #include "fiber.h"
+#include "trigger.h"
 #include <tarantool_ev.h>
 
 /*
@@ -122,6 +123,7 @@ static int
 coio_on_stop(void *data)
 {
        (void) data;
+       trigger_free_in_thread();
        cord_destroy(cord());
        free(cord());
        return 0;

```

</details>

Reply:
AFAICS the only place where we call `cord_destroy` from another thread is `cord_join`, but it should be okay to move `cord_destroy` to `cord_thread_func` from there. I'd do it and add an assertion to `cord_destroy` that it's called from the destroyed cord.

Reply:
`cord_destroy` clears cord's slab cache, so `cord_join` can't check `fiber->f_ret` and `fiber->diag` for the joined cord's fiber, because the latter is already cleared.
I would still have to split `cord_destroy` in two for the sake of `cord_join`. The first part would perform the thread-specific deinitialization, and the second one would destroy the cord's slab cache. This doesn't look much simpler than calling `trigger_free_in_thread` in all the needed places manually IMO.
I haven't done it yet, asked for your opinion in chat.

Reply:
I don't like that you call `trigger_free_in_thread` here and there directly. First, this introduces a dependency from `trigger.h`. Second, if we want to do some extra destruction work at thread exit, we'll have to patch all those places where you add `trigger_free_in_thread` now. Let's add a wrapper for this - `cord_exit`.